lib/ansible/modules/cloud/amazon/iam_policy.py

```diff
@@ -112,240 +114,215 @@
+                pdoc = json.dumps(json.load(json_data), sort_keys=True)
+                json_data.close()
+        except IOError as e:
+            if e.errno == 2:
+                raise PolicyError('policy_document {0:!r} does not exist'.format(self.policy_document))
+        return pdoc
 
+    def get_policy_from_json(self):
+        try:
+            if isinstance(self.policy_json, string_types):
+                pdoc = json.dumps(json.loads(self.policy_json), sort_keys=True)
+            else:
+                pdoc = json.dumps(self.policy_json, sort_keys=True)
+        except Exception as e:
+            raise PolicyError('Failed to convert the policy into valid JSON: %s' % str(e))
+        return pdoc
 
-def group_action(module, iam, name, policy_name, skip, pdoc, state):
-    policy_match = False
-    changed = False
-    msg = ''
-    try:
-        current_policies = [cp for cp in iam.get_all_group_policies(name).
-                            list_group_policies_result.
-                            policy_names]
+    def create(self):
         matching_policies = []
-        for pol in current_policies:
-            if urllib.parse.unquote(iam.get_group_policy(name, pol).
-                                    get_group_policy_result.policy_document) == pdoc:
-                policy_match = True
+        policy_doc = self.get_policy_text()
+        for pol in self.list():
+            if self.get(pol) == policy_doc:
                 matching_policies.append(pol)
-                msg = ("The policy document you specified already exists "
-                       "under the name %s." % pol)
-        if state == 'present':
-            # If policy document does not already exist (either it's changed
-            # or the policy is not present) or if we're not skipping dupes then
-            # make the put call.  Note that the put call does a create or update.
-            if not policy_match or (not skip and policy_name not in matching_policies):
-                changed = True
-                iam.put_group_policy(name, policy_name, pdoc)
-        elif state == 'absent':
-            try:
-                iam.delete_group_policy(name, policy_name)
-                changed = True
-            except boto.exception.BotoServerError as err:
-                error_msg = boto_exception(err)
-                if 'cannot be found.' in error_msg:
-                    changed = False
-                    module.exit_json(changed=changed,
-                                     msg="%s policy is already absent" % policy_name)
 
-        updated_policies = [cp for cp in iam.get_all_group_policies(name).
-                            list_group_policies_result.
-                            policy_names]
-    except boto.exception.BotoServerError as err:
-        error_msg = boto_exception(err)
-        module.fail_json(changed=changed, msg=error_msg)
+        if self.policy_name not in matching_policies or not self.skip_duplicates:
```

Comment:
It looks like this broke skip_duplicates. The module used to use the policy document as the unique identifier rather than the name. It used to mean you could do this:
```
  - name: create testpolicy1 (changed is true)
    iam_policy:
      iam_type: group
      iam_name: testgroup
      policy_name: testpolicy1
      state: present
      policy_document: policy.json

  - name: do not create testpolicy2 because testpolicy1 already matches the policy (changed is false)
    iam_policy:
      iam_type: group
      iam_name: testgroup
      policy_name: testpolicy2
      state: present
      policy_document: policy.json

  - name: use skip_duplicates to create testpolicy2 despite testpolicy1 already matching the policy (changed is true)
    iam_policy:
      iam_type: group
      iam_name: testgroup
      policy_name: testpolicy2
      state: present
      policy_document: policy.json
      skip_duplicates: false
```

It looks like there's a gap in the integration tests since they didn't catch that.

Reply:
I really wish we could deprecate that behaviour (at least turning it off by default) it's really confusing if you're not looking for it.

Reply:
That's definitely an option. We could remove or disable by default in 4 releases if we fix this and add a deprecation warning.

Reply:
@s-hertel FYI I've reworked the test suite in #62955 to make sure we test the current expected behaviour and test a few other things more thoroughly.